### PR TITLE
Upgrade to Google Play's new subscription model and APIs

### PR DESCRIPTION
**Pull request by tomwadeson**
## What does this change?

This PR migrates us to Google Play's new subscription model (and associated APIs.) See here for more details:  https://developer.android.com/google/play/billing/compatibility

### Outstanding

- [ ] Prepare for the removal of `subscriptionId`s from the Real Time Developer Notifications https://developer.android.com/google/play/billing/compatibility#real-time
- [ ] Update the cache in the case of a cache miss on the Google subscription status endpoint

## How to test

This is a good question!

There is an existing absence of test coverage for the affected area of the codebase due, I suspect, to the difficulty (impossibility?) of automating any useful/non-trivial integration tests:

To test this manually requires access to the Google Play project; a user account that has license testing privileges (and potentially in the right state—i.e., having not already taken out a subscription, or having not previously benefited from a free trial); and a debug build of the Android app that is configured to surface the underlying `purchaseToken`. Testing the lifecycle of subscriptions also has slow feedback loops as we must wait for subscriptions to progress through their various states of activation, renewal and expiry etc. without having the means to control this programmatically. 

We have conducted _many_ manual tests of this migration effort (and refined the implementation as a result of edge cases that we've encountered), but I suspect that to be entirely confident of this change prior to pushing it live, we would want to run it in parallel with the existing implementation to compare (hopefully) like-for-like.

## Accessibility

-   [ ] [Tested with screen reader](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#screen-reader)
-   [ ] [Navigable with keyboard](https://github.com/guardian/accessibility/blob/main/people-and-technology/02-physical.md#keyboard)
-   [ ] [Colour contrast passed](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#contrast)
-   [ ] [The change doesn't use only colour to convey meaning](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#use-of-colour)

### CODE REVIEW

**Review by tjmw — APPROVED** [review]
This looks really good! I've left a handful of comments but nothing major - all nitpicky stuff to take or leave.

File: typescript/src/services/google-play.ts [hunk 1]
```diff
@@ -1,28 +1,165 @@
...
+
+        const purchase =
+            await client.purchases.subscriptionsv2.get({ packageName: packageName, token: purchaseToken })
+
+        // A subscription purchase refers to one or many underlying products ("line items".) However, by convention (and by
+        // constraining/controling the UX within the app), we will always assume that a subscription purchase refers to exactly
+        // one product.
+        const product =
+            purchase.data.lineItems?.find(x => x !== undefined) ??
+            throwError("The subscription purchase does not refer to a product")
```
[the remark this answers is not in the document]
Comment (tomwadeson) [reply typescript/src/services/google-play.ts:46]: Thanks, that's a good catch!

File: typescript/src/services/google-play.ts [hunk 2]
```diff
@@ -1,28 +1,165 @@
...
+            await client.monetization.subscriptions.get({ packageName: packageName, productId: productId } )
+
+        const basePlan =
+            subscription.data.basePlans?.find(x => x.basePlanId == basePlanId) ??
+            throwError("Unable to determine the base plan for the product")
+
+        const billingPeriodDuration =
+            basePlan.autoRenewingBasePlanType?.billingPeriodDuration ??
+            basePlan.prepaidBasePlanType?.billingPeriodDuration ??
+            throwError("Unable to determine a billing period duration for the base plan")
```
[the remark this answers is not in the document]
Comment (tomwadeson) [reply typescript/src/services/google-play.ts:79]: I'm unable to inline it in this case as `??` expects an expression while `throw` is a statement.

But maybe this is a signal for me that I should leave my expression bias behind in TS, and it'd be more idiomatic to do something like:

```typescript
let billingPeriodDuration = null
if (basePlan.autoRenewingBasePlanType?.billingPeriodDuration) {
    billingPeriodDuration = ...
} else if (basePlan.prepaidBasePlanType?.billingPeriodDuration) {
    billingPeriodDuration = ...
} else {
    throw Error(...
}
```

Very happy to be guided by your tastes on this!

File: typescript/src/services/google-play.ts [hunk 3]
```diff
@@ -1,28 +1,165 @@
...
+export async function fetchGoogleSubscription(
+    purchaseToken: string,
+    packageName: string
+): Promise<GoogleSubscription> {
+    try {
+        const client =
+            await initialiseAndroidPublisherClient()
+
+        const purchase =
+            await client.purchases.subscriptionsv2.get({ packageName: packageName, token: purchaseToken })
```
[the remark this answers is not in the document]
Comment (tomwadeson) [reply typescript/src/services/google-play.ts:39]: Thanks for that hint. I'll roll that out everywhere I can (but might leave `purchaseToken` named as it is in this case, just to be consistent with Google's names for things, even if they're being a bit inconsistent here!) Much tidier 👍

File: typescript/src/services/google-play.ts [hunk 4]
```diff
@@ -1,28 +1,165 @@
...
+        return {
+            startTime: parseNullableDate(startTime),
+            expiryTime: new Date(expiryTime),
+            userCancellationTime: parseNullableDate(userCancellationTime),
+            autoRenewing: autoRenewing,
+            productId: productId,
+            billingPeriodDuration: billingPeriodDuration,
+            freeTrial: isFreeTrial(offerId, latestOrderId)
+        }
+    } catch (error: any) {
```
[the remark this answers is not in the document]
Comment (tomwadeson) [reply typescript/src/services/google-play.ts:97]: I get: `Catch clause variable type annotation must be 'any' or 'unknown' if specified.ts(1196)` for that. But I guess the explicit ascription isn't doing anything, so I'll remove that anyway!

File: typescript/src/services/google-play.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (tjmw) [reply typescript/src/services/google-play.ts:79]: Ah right, that makes sense, apologies for the derail in that case.

Maybe a pattern which is a hybrid of the two approaches would be nice, and we could avoid the let/mutation:

```js
const billingPeriodDuration =
    basePlan.autoRenewingBasePlanType?.billingPeriodDuration ??
    basePlan.prepaidBasePlanType?.billingPeriodDuration;

if (!billingPeriodDuration) {
    throw new Error("Unable to determine a billing period duration for the base plan");
}
```

[the remark this answers is not in the document]
Comment (tjmw) [reply typescript/src/services/google-play.ts:79]: Or just leave it as is!

[the remark this answers is not in the document]
Comment (tomwadeson) [reply typescript/src/services/google-play.ts:79]: Ah yes, that looks like a good solution, cheers!
